Lib/importlib/_bootstrap_external.py

```diff
@@ -1711,6 +1711,78 @@ def __repr__(self):
+    app must be .dylib objects, contained in a framework with appropriate
+    metadata, stored in the ``Frameworks`` folder of the packaged app. There
+    can be only a single binary per framework, and there can be no executable
+    binary material outside the Frameworks folder.
+
+    If you're trying to run ``from foo.bar import _whiz``, and ``_whiz`` is
+    implemented with the binary module ``foo/bar/_whiz.abi3.dylib`` (or any
+    other ABI .dylib extension), this loader will look for
+    ``{sys.executable}/Frameworks/foo.bar._whiz.framework/_whiz.abi3.dylib``
+    (forming the package name by taking the full import path of the library,
+    and replacing ``/`` with ``.``).
+
+    However, the ``__file__`` attribute of the ``_whiz`` module will report as
+    the original location inside the ``foo/bar`` subdirectory. This so that
+    code that depends on walking directory trees will continue to work as
+    expected based on the *original* file location.
+
+    The Xcode project building the app is responsible for converting any
+    ``.dylib`` files from wherever they exist in the ``PYTHONPATH`` into
+    frameworks in the ``Frameworks`` folder (including the addition of
+    framework metadata, and signing the resulting framework). This will usually
+    be done with a build step in the Xcode project; see the iOS documentation
+    for details on how to construct this build step.
+    """
+    def __init__(self, fullname, dylib_file, parent_paths=None):
+        super().__init__(fullname, dylib_file)
+        self.parent_paths = parent_paths
+
+    def create_module(self, spec):
+        mod = super().create_module(spec)
+        if self.parent_paths:
+            for parent_path in self.parent_paths:
+                if _path_isdir(parent_path):
+                    mod.__file__ = _path_join(
+                        parent_path,
+                        _path_split(self.path)[-1],
+                    )
+                    break
+        return mod
+
+
+class AppleFrameworkFinder:
+    """A finder for modules that have been packaged as Apple Frameworks
+    for compatibility with Apple's App Store policies.
+
+    See AppleFrameworkLoader for details.
+    """
+    def __init__(self, frameworks_path):
+        self.frameworks_path = frameworks_path
+
+    def find_spec(self, fullname, paths, target=None):
+        name = fullname.split(".")[-1]
+
+        for extension in EXTENSION_SUFFIXES:
+            dylib_file = _path_join(
+                self.frameworks_path,
+                f"{fullname}.framework", f"{name}{extension}"
+            )
+            _bootstrap._verbose_message("Looking for Apple Framework dylib {}", dylib_file)
+
+            dylib_exists = _path_isfile(dylib_file)
+            if dylib_exists:
+                loader = AppleFrameworkLoader(fullname, dylib_file, paths)
+                return _bootstrap.spec_from_loader(fullname, loader)
```

Comment:
If you take advantage of `spec.loader_state`[^1] here then you don't need a custom `__init__()` on the loader.  This involves setting `spec.loader_state.path` instead of passing it as an argument to the loader.

[^1]: For an example, see `FrozenImporter` in Lib/importlib/_bootstrap.py.

```suggestion
            if dylib_exists:
                loader = AppleFrameworkLoader(fullname, dylib_file)
                spec = _bootstrap.spec_from_loader(fullname, loader)
                spec.loader_state = type(sys.implementation)(
                    parent_paths=path,
                )
                return spec
```

Reply:
Furthermore, I recommend doing as much as possible in the finder.  Rather than passing `paths` and calculating `__file__` in the loader, why not calculate it here and pass that?  That definitely would make the loader simpler.

```suggestion
            if dylib_exists:
                origfile = None
                if paths:
                    for parent_path in paths:
                        if _path_isdir(parent_path):
                            origfile = _path_join(
                                parent_path,
                                _path_split(self.path)[-1],
                            )
                            break
                loader = AppleFrameworkLoader(fullname, dylib_file)
                spec = _bootstrap.spec_from_loader(fullname, loader)
                spec.loader_state = type(sys.implementation)(
                    origfile=origfile,
                )
                return spec
```

The only caveat is that doing it in the loader is necessary if you expect that the calculated `__file__` might be different if done later (e.g. when reloading the module).  However, in this case you have been doing it in the loader's `create_module()` which means it would only happen when the module is first loaded and not when reloaded.  Thus I'd expect you'd be fine calculating `__file__` in the finder.

Reply:
The path definitely can't change after initial load; optimising in the finder makes sense.